apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.jsx

```diff
@@ -0,0 +1,203 @@
+import ProgressTableLevelBubble from './ProgressTableLevelBubble';
+import * as progressStyles from '@cdo/apps/templates/progress/progressStyles';
+import color from '@cdo/apps/util/color';
+import firehoseClient from '@cdo/apps/lib/util/firehose';
+import _ from 'lodash';
+import {LevelStatus} from '@cdo/apps/util/sharedConstants';
+
+const UNPLUGGED_BUBBLE_WIDTH = getUnpluggedWidth();
+
+const styles = {
+  container: {
+    ...progressStyles.flexBetween,
+    position: 'relative',
+    whiteSpace: 'nowrap'
+  },
+  background: {
+    height: 10,
+    backgroundColor: color.lighter_gray,
+    position: 'absolute',
+    left: 10,
+    right: 10
+  },
+  sublevelContainer: {
+    position: 'relative',
+    display: 'inline-block'
+  }
+};
+export default class ProgressTableDetailCell extends React.Component {
+  static widthForLevels(levels) {
+    return (
+      levels.reduce((sum, level) => {
+        return (
+          sum +
+          (level.isUnplugged
+            ? UNPLUGGED_BUBBLE_WIDTH
+            : progressStyles.BUBBLE_CONTAINER_WIDTH) +
+          (level.sublevels || []).length *
+            progressStyles.LETTER_BUBBLE_CONTAINER_WIDTH
+        );
+      }, 0) +
+      2 * progressStyles.CELL_PADDING
+    );
+  }
+
+  static propTypes = {
+    studentId: PropTypes.number.isRequired,
+    sectionId: PropTypes.number.isRequired,
+    levels: PropTypes.arrayOf(levelTypeWithoutStatus).isRequired,
+    studentProgress: PropTypes.objectOf(studentLevelProgressType).isRequired,
+    stageExtrasEnabled: PropTypes.bool
+  };
+
+  constructor(props) {
+    super(props);
+    this.recordBubbleClick = this.recordBubbleClick.bind(this);
+  }
+
+  shouldComponentUpdate(nextProps) {
+    return !_.isEqual(this.props, nextProps);
+  }
+
+  recordBubbleClick(levelId) {
+    firehoseClient.putRecord(
+      {
+        study: 'teacher_dashboard_actions',
+        study_group: 'progress',
+        event: 'go_to_level',
+        data_json: JSON.stringify({
+          student_id: this.props.studentId,
+          section_id: this.props.sectionId,
+          level_id: levelId
+        })
+      },
+      {includeUserId: true}
+    );
+  }
+
+  buildBubbleUrl(level) {
+    if (!level.url) {
+      return null;
+    }
+    const {studentId, sectionId} = this.props;
+    return `${level.url}?section_id=${sectionId}&user_id=${studentId}`;
+  }
+
+  renderSublevels(level) {
+    return (
+      <div>
+        {level.sublevels.map(sublevel => {
+          const subProgress = this.props.studentProgress[sublevel.id];
+          const subStatus = subProgress && subProgress.status;
+          return (
+            <div
+              key={sublevel.id}
+              style={styles.sublevelContainer}
+              onClick={_ => this.recordBubbleClick(sublevel.id)}
+            >
+              <ProgressTableLevelBubble
+                levelStatus={subStatus}
+                disabled={!!level.bonus && !this.props.stageExtrasEnabled}
```

Comment:
What type is level.bonus?  (Wondering why we need the `!!` in front of it.)

Reply:
it's an optional bool, hence the need for `!!`.